File: examples/typescript/src/patterns/parallel/parallel-step-flows.ts
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Channel,
  StepList,
  StepMovement,
  Wait,
  deadEnd,
  doubleCodec,
  goToMany,
  gracefulComplete,
  stringCodec,
  voidCodec,
  withCancelingSiblingSteps,
  type Context,
  type Flow,
  type PersistenceSchema,
  type Step,
  type StepDecision,
} from "@superdurable/dex";

class WorkA implements Step<string> {
  public readonly inputCodec = stringCodec;
  public getStepType(): string {
    return "WorkA";
  }
  public execute(_context: Context, input: string): StepDecision {
    return gracefulComplete(`A:${input}`);
  }
}

class WorkB implements Step<string> {
  public readonly inputCodec = stringCodec;
  public getStepType(): string {
    return "WorkB";
  }
  public execute(_context: Context, input: string): StepDecision {
    return gracefulComplete(`B:${input}`);
  }
}

class StaticInit implements Step<string> {
  public readonly inputCodec = stringCodec;
  public getStepType(): string {
    return "Init";
  }
  public execute(_context: Context, input: string): StepDecision {
    return goToMany(
      StepMovement.of(WorkA, input),
      StepMovement.of(WorkB, input),
    );
  }
}

export class StaticParallelStepsFlow implements Flow<string> {
  private readonly init = new StaticInit();
  private readonly workA = new WorkA();
  private readonly workB = new WorkB();
  public getFlowType(): string {
    return "StaticParallelStepsFlow";
  }
  public getSteps() {
    return StepList.startStep(this.init).otherSteps(this.workA, this.workB);
  }
  public getPersistenceSchema(): PersistenceSchema {
    return {};
  }
}

class DynamicDoWork implements Step<number> {
  public readonly inputCodec = doubleCodec;
  public getStepType(): string {
    return "DoWork";
  }
  public async execute(_context: Context, input: number): Promise<StepDecision> {
    await new Promise<void>((resolve) => {
      setTimeout(resolve, 50 + Math.floor(Math.random() * 450));
    });
    return gracefulComplete(input);
  }
}

class DynamicInit implements Step<number> {
  public readonly inputCodec = doubleCodec;
  public getStepType(): string {
    return "Init";
  }
  public execute(_context: Context, count: number): StepDecision {
    return goToMany(
      ...Array.from({ length: count }, (_, index) =>
        StepMovement.of(DynamicDoWork, index),
      ),
    );
  }
}

export class DynamicParallelStepsFlow implements Flow<number> {
  private readonly init = new DynamicInit();
  private readonly work = new DynamicDoWork();
  public getFlowType(): string {
    return "DynamicParallelStepsFlow";
  }
  public getSteps() {
    return StepList.startStep(this.init).otherSteps(this.work);
  }
  public getPersistenceSchema(): PersistenceSchema {
    return {};
  }
}

const completeCh = new Channel("parallel-complete", voidCodec);

class AwaitDoWork implements Step<number> {
  public readonly inputCodec = doubleCodec;
  public getStepType(): string {
    return "DoWork";
  }
  public async execute(context: Context, _input: number): Promise<StepDecision> {
    await new Promise<void>((resolve) => {
      setTimeout(resolve, 50 + Math.floor(Math.random() * 450));
    });
    completeCh.publish(context, undefined);
    return deadEnd();
  }
}

class Await implements Step<number> {
  public readonly inputCodec = doubleCodec;
  public getStepType(): string {
    return "Await";
  }
  public waitFor(_context: Context, count: number): Wait {
    return Wait.until(completeCh.forN(count));
  }
  public execute(_context: Context, count: number): StepDecision {
    return gracefulComplete(count);
  }
}

class AwaitInit implements Step<number> {
  public readonly inputCodec = doubleCodec;
  public getStepType(): string {
    return "Init";
  }
  public execute(_context: Context, count: number): StepDecision {
    return goToMany(
      StepMovement.of(Await, count),
      ...Array.from({ length: count }, (_, index) =>
        StepMovement.of(AwaitDoWork, index),
      ),
    );
  }
}

export class AwaitParallelStepsFlow implements Flow<number> {
  private readonly init = new AwaitInit();
  private readonly work = new AwaitDoWork();
  private readonly awaitStep = new Await();
  public getFlowType(): string {
    return "AwaitParallelStepsFlow";
  }
  public getSteps() {
    return StepList.startStep(this.init).otherSteps(this.work, this.awaitStep);
  }
  public getPersistenceSchema(): PersistenceSchema {
    return { channels: [completeCh] };
  }
}

class FirstWinDoWork implements Step<number> {
  public readonly inputCodec = doubleCodec;
  public getStepType(): string {
    return "DoWork";
  }
  public async execute(_context: Context, input: number): Promise<StepDecision> {
    await new Promise<void>((resolve) => {
      setTimeout(resolve, 50 + Math.floor(Math.random() * 450));
    });
    return withCancelingSiblingSteps(gracefulComplete(input), FirstWinDoWork);
  }
}

class FirstWinInit implements Step<number> {
  public readonly inputCodec = doubleCodec;
  public getStepType(): string {
    return "Init";
  }
  public execute(_context: Context, count: number): StepDecision {
    return goToMany(
      ...Array.from({ length: count }, (_, index) =>
        StepMovement.of(FirstWinDoWork, index),
      ),
    );
  }
}

export class FirstWinParallelStepsFlow implements Flow<number> {
  private readonly init = new FirstWinInit();
  private readonly work = new FirstWinDoWork();
  public getFlowType(): string {
    return "FirstWinParallelStepsFlow";
  }
  public getSteps() {
    return StepList.startStep(this.init).otherSteps(this.work);
  }
  public getPersistenceSchema(): PersistenceSchema {
    return {};
  }
}

export const staticParallelStepsFlow = new StaticParallelStepsFlow();
export const dynamicParallelStepsFlow = new DynamicParallelStepsFlow();
export const awaitParallelStepsFlow = new AwaitParallelStepsFlow();
export const firstWinParallelStepsFlow = new FirstWinParallelStepsFlow();

File: examples/rust/src/patterns/parallel/parallel_step_flows.rs
// Copyright (c) 2022-2026 Super Durable, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use dex_sdk::{
    Channel, Context, Flow, HandlerResult, PersistenceSchema, Step, StepDecision, StepList,
    StepMovement, Wait,
};
use std::{
    sync::LazyLock,
    thread,
    time::{Duration, SystemTime},
};

#[derive(Default)]
pub struct StaticParallelStepsFlow {
    init: StaticInit,
    work_a: WorkA,
    work_b: WorkB,
}
impl Flow for StaticParallelStepsFlow {
    type StartInput = String;
    fn steps(&self) -> StepList<'_, String> {
        StepList::start(&self.init)
            .and(&self.work_a)
            .and(&self.work_b)
    }
}
#[derive(Default)]
struct StaticInit;
impl Step for StaticInit {
    type Input = String;
    fn step_type(&self) -> &'static str {
        "InitStep"
    }
    fn execute(&self, _: &mut Context, input: String) -> HandlerResult<StepDecision> {
        Ok(StepDecision::go_to_many([
            StepMovement::to(&WorkA, input.clone()),
            StepMovement::to(&WorkB, input),
        ]))
    }
}
#[derive(Default)]
struct WorkA;
impl Step for WorkA {
    type Input = String;
    fn step_type(&self) -> &'static str {
        "WorkAStep"
    }
    fn execute(&self, _: &mut Context, input: String) -> HandlerResult<StepDecision> {
        Ok(StepDecision::graceful_complete(format!("A:{input}")))
    }
}
#[derive(Default)]
struct WorkB;
impl Step for WorkB {
    type Input = String;
    fn step_type(&self) -> &'static str {
        "WorkBStep"
    }
    fn execute(&self, _: &mut Context, input: String) -> HandlerResult<StepDecision> {
        Ok(StepDecision::graceful_complete(format!("B:{input}")))
    }
}

#[derive(Default)]
pub struct DynamicParallelStepsFlow {
    init: DynamicInit,
    work: DynamicWork,
}
impl Flow for DynamicParallelStepsFlow {
    type StartInput = usize;
    fn steps(&self) -> StepList<'_, usize> {
        StepList::start(&self.init).and(&self.work)
    }
}
#[derive(Default)]
struct DynamicInit;
impl Step for DynamicInit {
    type Input = usize;
    fn step_type(&self) -> &'static str {
        "InitStep"
    }
    fn execute(&self, _: &mut Context, count: usize) -> HandlerResult<StepDecision> {
        Ok(StepDecision::go_to_many(
            (0..count).map(|index| StepMovement::to(&DynamicWork, index)),
        ))
    }
}
#[derive(Default)]
struct DynamicWork;
impl Step for DynamicWork {
    type Input = usize;
    fn step_type(&self) -> &'static str {
        "DoWorkStep"
    }
    fn execute(&self, _: &mut Context, input: usize) -> HandlerResult<StepDecision> {
        random_delay();
        Ok(StepDecision::graceful_complete(input))
    }
}

#[derive(Default)]
pub struct AwaitParallelStepsFlow {
    init: AwaitInit,
    work: AwaitWork,
    await_step: AwaitStep,
}
impl Flow for AwaitParallelStepsFlow {
    type StartInput = usize;
    fn steps(&self) -> StepList<'_, usize> {
        StepList::start(&self.init)
            .and(&self.work)
            .and(&self.await_step)
    }
    fn persistence(&self) -> PersistenceSchema {
        PersistenceSchema::new().channel(&COMPLETE)
    }
}
#[derive(Default)]
struct AwaitInit;
impl Step for AwaitInit {
    type Input = usize;
    fn step_type(&self) -> &'static str {
        "InitStep"
    }
    fn execute(&self, _: &mut Context, count: usize) -> HandlerResult<StepDecision> {
        let mut movements = vec![StepMovement::to(&AwaitStep, count)];
        movements.extend((0..count).map(|index| StepMovement::to(&AwaitWork, index)));
        Ok(StepDecision::go_to_many(movements))
    }
}
#[derive(Default)]
struct AwaitWork;
impl Step for AwaitWork {
    type Input = usize;
    fn step_type(&self) -> &'static str {
        "DoWorkStep"
    }
    fn execute(&self, context: &mut Context, _: usize) -> HandlerResult<StepDecision> {
        random_delay();
        COMPLETE.publish(context, ())?;
        Ok(StepDecision::dead_end())
    }
}
#[derive(Default)]
struct AwaitStep;
impl Step for AwaitStep {
    type Input = usize;
    fn step_type(&self) -> &'static str {
        "AwaitStep"
    }
    fn wait_for(&self, _: &mut Context, count: usize) -> HandlerResult<Wait> {
        Ok(Wait::until(COMPLETE.for_n(count)))
    }
    fn execute(&self, _: &mut Context, count: usize) -> HandlerResult<StepDecision> {
        Ok(StepDecision::graceful_complete(count))
    }
}
static COMPLETE: LazyLock<Channel<()>> = LazyLock::new(|| Channel::new("parallel-complete"));

#[derive(Default)]
pub struct FirstWinParallelStepsFlow {
    init: FirstWinInit,
    work: FirstWinWork,
}
impl Flow for FirstWinParallelStepsFlow {
    type StartInput = usize;
    fn steps(&self) -> StepList<'_, usize> {
        StepList::start(&self.init).and(&self.work)
    }
}
#[derive(Default)]
struct FirstWinInit;
impl Step for FirstWinInit {
    type Input = usize;
    fn step_type(&self) -> &'static str {
        "InitStep"
    }
    fn execute(&self, _: &mut Context, count: usize) -> HandlerResult<StepDecision> {
        Ok(StepDecision::go_to_many(
            (0..count).map(|index| StepMovement::to(&FirstWinWork, index)),
        ))
    }
}
#[derive(Default)]
struct FirstWinWork;
impl Step for FirstWinWork {
    type Input = usize;
    fn step_type(&self) -> &'static str {
        "DoWorkStep"
    }
    fn execute(&self, _: &mut Context, input: usize) -> HandlerResult<StepDecision> {
        random_delay();
        Ok(StepDecision::graceful_complete(input).cancel_sibling_step(&FirstWinWork))
    }
}

fn random_delay() {
    let nanos = SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .expect("system clock must follow Unix epoch")
        .subsec_nanos();
    thread::sleep(Duration::from_millis(50 + u64::from(nanos % 450)));
}
